### PR TITLE
Add HTTPS support

### PR DESCRIPTION
**Pull request by Jakski**
Now configuration has separate section for connection specific settings. I plan to add HTTP 2 HTTPS redirection specific settings, so having things separated may become handy, because this way we will be able to pass all connection's settings in one object from config to HapiJS's server.connection() method. It adds some boilerplate, so I would like to read your opinion on this.

### CODE REVIEW

**Comment by arkadiusz-wieczorek** [comment]
+1

**Comment by kuba-orlik** [comment]
I've reviewed the changes and I don't see https mentioned explicitly anywhere. Does this PR add https support? If so, did you test it? If so, how? Please elaborate a little bit :)

**Comment by Jakski** [comment]
I enabled developer to pass parameters directly to [server.connection()](http://hapijs.com/api#serverconnectionoptions) method via ConfigManager, e.g. settings to enable TLS/SSL. You can take a look at [hello-world](https://github.com/Rayvenden/hello-world/tree/https_example) example to see how to use it. We can pass options object as described in [NodeJS docs](https://nodejs.org/api/https.html#https_https_createserver_options_requestlistener) to tell a server, we're gonna use TLS/SSL. Actually 'Add HTTPS support' maybe isn't the best title for this commit, but it's all about enabling developer to use HTTPS.

**Comment by kuba-orlik** [comment]
I've cloned the respository and checked out the `https_example` branch, but couldn't get it to work - firefox says "Unable to connect" when I use https. Could you provide step-by-step instructions to how can I get it to work? :)

**Comment by Jakski** [comment]
Okay. I've finished whole HTTPS feature together with redirections from HTTP, so let me explain it all from the start. Config has 2 sections now: one for connections and one for redirections. I've tried to do it somehow easier, but couldn't find a way, because I would like it to support at least these 3 sub-features:
- only HTTP usage
- HTTP and HTTPS parallel usage
- HTTP and HTTPS usage + HTTP to HTTPS redirection(force users to use secure connection)

### How `connections` section works?

It's just a container with all connections configs. It doesn't matter how you name your connection sub-section, because `www_server.start()` iterates through all of them. All sub-sections should be objects with properties prepared to be passed directly to [server.connection()](http://hapijs.com/api#serverconnectionoptions) method from HapiJS. As described in HapiJS docs, passing `tls` object in config causes connection to use TLS/SSL. I prepared example of how to do it in [hello-world](https://github.com/Rayvenden/hello-world/tree/https_example)(it's been also updated since my last comment).

### How `redirections` section works?

It's also a container for sub-sections which are being later iterated by `www_server.start()`, so sub-sections names have no meaning. Every sub-section should have following properties:
- `protocol`
- source redirection port(`from`)
- destination redirection port(`to`)

`www_server.start()` reads this sub-sections and tells server how to handle connection on each port by calling [server.ext()](http://hapijs.com/api#serverextevent-method-options) from HapiJS.

### How logging changed?

Sealious used to print information about opened connection after initialization. It wasn't multi-connection aware approach. Now it logs all opened connections based on their sub-sections in config. Error handling on socket bind has been fixed as well, so go on and try it :)

### How to test HTTPS?

After getting [hello-world](https://github.com/Rayvenden/hello-world/tree/https_example) and [www_server](https://github.com/Sealious/sealious-www-server/tree/trello%23https_support) and launching it(`sudo` isn't needed) following links should get you to HTTPS version of service on 4430 port(HTTP 2 HTTPS redirection is enabled in example):
- https://localhost:4430/
- http://localhost:8080/

Your browser will warn you about insecure certificate, because I self-signed it ;)
I think we will have to discuss it IRL, but congratulations if you've finished reading.

**Comment by kuba-orlik** [comment]
Great job! I cloned the repo, followed your directions and managed to get https to work on my local machine. 

Kudos for exhaustive description of your changes and improvements to the Logger output.

Two things to note:
1. >  It doesn't matter how you name your connection sub-section, because www_server.start() iterates through all of them.
   > 
   > It's also a container for sub-sections which are being later iterated by www_server.start(), so sub-sections names have no meaning.
   
   Wouldn't it make more sense to use an array here? The config section could be a little thinner visually thanks to that, and the developer wouldn't have to come up with meaningless strings:
   
   ```
     "connections": [{
         "port": 4430,
         "tls": {
             "key": "fs.readFileSync(\"sealious.key\")",
             "cert": "fs.readFileSync(\"sealious.crt\")"
         }
     }]
   ```
2. We need to make the `www-server` plugin less... _mysterious_. Please use the description you provided us with above to make a nice `README.md` for the plugin and include it in the PR :)

> congratulations if you've finished reading.

Thank you again for the extensive description :)

**Comment by Jakski** [comment]
Good notes. I will change config structure and add docs soon.

**Comment by Jakski** [comment]
Cannot use arrays. I've tried it, but it turns out, that `merge` module overwrites arrays in default configuration with ones from user configuration. If you have HTTP default config and try to add HTTPS from `index.js`, only HTTPS work, because all `connections` sections is being overwritten.

**Comment by kuba-orlik** [comment]
Isn't that the desired result?

**Comment by Jakski** [comment]
I just realized, it actually makes configuration more intuitive :D .  Will just have to change `hello_world`, so it opens HTTP connection explicitly in `index.js`.

File: README.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+WWW-server configuration allows you to open ports for listening, define
+protocols served on those ports and redirections(useful when you want to
+force users to use HTTPS by redirecting them from HTTP). Configuration is
+separated into 2 sections: `connections` and `redirections`.
+
+### How `connections` section works?
+It's just an array with all connections configurations. All sub-sections should
+be objects with properties prepared to be passed directly to [server.connection()](http://hapijs.com/api#serverconnectionoptions) method from HapiJS.
+As described in HapiJS docs, passing `tls` object in config causes connection to use TLS/SSL.
+You can take a look at example of how to do it in [hello-world](https://github.com/Rayvenden/hello-world/tree/https_example).
```
Comment (kuba-orlik) [inline README.md]: I think we'd need an example configuration here! :)

**Comment by Rayvenden** [comment]
Will extensive commenting [hello-world](https://github.com/Rayvenden/hello-world/tree/https_example) be enough?(BTW: `hello-world` is still my private repository. I will push it into official, if you provide me privileges :P )

**Comment by kuba-orlik** [comment]
Hi, I'd like to keep hello-world focused more on Sealious' structure,
rather than  on a specific channel. That's why I think it'd be best to keep
the www-server configuration examples within the www-server Readme :)

**Comment by Jakski** [comment]
I've added some examples. Can you review it?

**Comment by kuba-orlik** [comment]
Merged to `alpha` in https://github.com/Sealious/sealious-www-server/commit/1643b3deaf17efb00ed603d3d82b75728a8eb619

**Pull request closed without merging**